concent_api/core/payments/sci_callback.py

```diff
@@ -0,0 +1,260 @@
+        to       = transaction.to,
+        value    = transaction.value,
+        data     = transaction.data,
+    )
+
+    setattr(
+        transaction_signing_request,
+        'from',
+        generate_ethereum_address_from_ethereum_public_key_bytes(settings.CONCENT_ETHEREUM_PUBLIC_KEY),
+    )
+
+    return transaction_signing_request
+
+
+def send_request_to_middleman(middleman_message: GolemMessageFrame) -> bytes:
+    """
+    Opens socket connection to Middleman, sends Frame to MiddleMan through MiddleMan Protocol and receive response.
+
+    Returns raw Frame as bytes.
+    """
+
+    assert isinstance(middleman_message, GolemMessageFrame)
+
+    # Establish a TCP connection with Middleman.
+    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as client_socket:
+        client_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, True)
+        client_socket.connect((settings.MIDDLEMAN_ADDRESS, settings.MIDDLEMAN_PORT))
+        client_socket.settimeout(SCI_CALLBACK_MAXIMUM_TIMEOUT)
+
+        # Send the TransactionSigningRequest.
+        try:
+            send_over_stream(
+                connection=client_socket,
+                raw_message=middleman_message,
+                private_key=settings.CONCENT_PRIVATE_KEY,
+            )
+
+            # Read the response and close the connection.
+            return next(unescape_stream(connection=client_socket))
+        # Connection with Middleman times out before a complete response is received.
+        except socket.timeout as exception:
+            raise SCICallbackTimeoutError() from exception
+
+
+def deserialize_response_and_handle_errors(raw_response: bytes, request_id: int) -> SignedTransaction:
+    """
+    Deserialize received Frame and its payload and handle related errors.
+
+    Returns SignedTransaction message.
+    """
+
+    assert isinstance(raw_response, bytes)
+    assert isinstance(request_id, int)
+
+    try:
+        deserialized_message = GolemMessageFrame.deserialize(
+            raw_message=raw_response,
+            public_key=settings.CONCENT_PUBLIC_KEY,
+        )
+    # Received data frame is invalid or the signature does not match its content.
+    except MiddlemanProtocolError as exception:
+        raise SCICallbackFrameError() from exception
+    # If the received Golem message is malformed.
+    except MessageError as exception:
+        raise SCICallbackPayloadError() from exception
+
+    # Request ID in the received data frame does not match the one in the data frame that was sent.
+    if deserialized_message.request_id != request_id:
+        raise SCICallbackRequestIdError('MiddleMan response request_id does not match requested.')
+
+    # If the received frame is ErrorFrame.
+    if isinstance(deserialized_message, ErrorFrame):
+        raise SCICallbackPayloadError(
+            'Received frame is ErrorFrame.'
+        )
+
+    # If the received Golem message is TransactionRejected.
```

Comment:
this comment seems unnecessary - the code is clear about itself ;)

Reply:
:+1: